bundler/lib/bundler/compact_index_client/cache.rb

```diff
@@ -7,114 +7,107 @@ class CompactIndexClient
     class Cache
       attr_reader :directory
 
-      def initialize(directory)
+      def initialize(directory, fetcher = nil)
         @directory = Pathname.new(directory).expand_path
-        info_roots.each {|dir| mkdir(dir) }
-        mkdir(info_etag_root)
+        @updater = Updater.new(fetcher) if fetcher
+        @mutex = Thread::Mutex.new
+        @endpoints = Set.new
+
+        SharedHelpers.filesystem_access(@directory) do
+          @info_root = mkdir("info")
+          @special_characters_info_root = mkdir("info-special-characters")
+          @info_etag_root = mkdir("info-etags")
+        end
       end
 
       def names
-        lines(names_path)
+        fetch("names", names_path, names_etag_path)
       end
 
-      def names_path
-        directory.join("names")
+      def versions
+        fetch("versions", versions_path, versions_etag_path)
       end
 
-      def names_etag_path
-        directory.join("names.etag")
-      end
+      def info(name, remote_checksum = nil)
+        path = info_path(name)
 
-      def versions
-        versions_by_name = Hash.new {|hash, key| hash[key] = [] }
-        info_checksums_by_name = {}
-
-        lines(versions_path).each do |line|
-          name, versions_string, info_checksum = line.split(" ", 3)
-          info_checksums_by_name[name] = info_checksum || ""
-          versions_string.split(",") do |version|
-            delete = version.delete_prefix!("-")
-            version = version.split("-", 2).unshift(name)
-            if delete
-              versions_by_name[name].delete(version)
-            else
-              versions_by_name[name] << version
-            end
-          end
+        unless remote_checksum
```

Comment:
Purely style/taste: previously we had two early returns for the "update skipped" cases, and then the "update performed" case followed. Now we have one early return for the first skipped case and then symmetric logic for the others. I would either keep both early returns as they were before, or remove all of them.